Optimization to put LHS operand in registers for WGMMA before elementwise ops

ggengnv opened the pull request:
Notes
* Locally, benchmarked on H100 HBM3 700W for a sample workload i8 -> bf16 on two shapes. Seeing **15-20%** perf gain.
* Prefetching into RF will be put into another PR
* Coalesce issue is fixed in the latest commits
* I largely reuse existing Ampere logic for SMEM to RF copy with minor modifications for Hopper. More changes might be needed if SMEM bank conflicts becomes an issue or if we can't use this for TMA 

Hopper has two kinds of WGMMAs, "SS" (both operands in shmem) and "RS" (LHS operand A in registers).
In cases where we apply elementwise operations on A before WGMMA, Triton previously will copy A from global memory (GMEM) into registers (RF), perform the elementwise ops, and then copy to shared memory (SMEM) to perform SS WGMMA.

This PR adds an optimization for the case above to use RS GEMM, with the benefit of SMEM pipelining, i.e. copy from GMEM to SMEM, and then from SMEM to RF to finally perform RS WGMMA. The copying from GMEM to SMEM is necessary for coalesced access and allows for pipelining with LDGSTS (and potentially TMA in the future).

ggengnv commented:
Fixed global coalesce issue and the PR should be ready for full review. Prefetch will be added in another PR seeing we're already seeing perf gains with the current changes

chsigg commented on lib/Dialect/TritonGPU/IR/Dialect.cpp:
```diff
@@ -1979,14 +1983,20 @@ SmallVector<int64_t> NvidiaMmaEncodingAttr::getMMAv2Rep(ArrayRef<int64_t> shape,
-    return getTotalElemsPerThread(shape, eltTy);
+    assert(opIdx == 0);
+    auto instrMNK = getInstrShape();
+    auto wpt = getWarpsPerCTA();
```
You can use `warpsPerCTAM` from just above.

ggengnv replied:
Addressed in new split PR - resolving. Doing the same for all other comments in this PR, as soon as they are resolved in the corresponding split PR.

chsigg commented on lib/Dialect/TritonGPU/IR/Dialect.cpp:
```diff
@@ -1979,14 +1983,20 @@ SmallVector<int64_t> NvidiaMmaEncodingAttr::getMMAv2Rep(ArrayRef<int64_t> shape,
+    auto wpt = getWarpsPerCTA();
+    int repM = ceil<unsigned>(shapePerCTA[0], instrMNK[0] * wpt[0]);
+    int repK = ceil<unsigned>(shapePerCTA[1], instrMNK[2]);
+    return 4 * kWidth * repM * repK;
```
This needs an explanatory comment.

chsigg commented on lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp:
```diff
@@ -15,6 +17,96 @@ namespace gpu {
+
+// Returns whether we can hoist DotOp Encoding through `op`.
+// Roughly, whether op is elementwise and thus threads don't need
+// to exchange elements. But some ops are not current supported even though
```
'currently'

chsigg commented on lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp:
```diff
@@ -15,6 +17,96 @@ namespace gpu {
+  // Only consider custom conversions or arith ops.
+  // TODO(jlebar): Is this too restrictive?
+  if (!isa<FpToFpOp, BitcastOp>(op) && !isPureUnaryInlineAsm(op) &&
+      op->getDialect()->getTypeID() != TypeID::get<arith::ArithDialect>())
```
I'm aware this code was only moved but `!isa<arith::ArithDialect>(op->getDialect())` would be better.

chsigg commented on lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp:
```diff
@@ -15,6 +17,96 @@ namespace gpu {
+bool canHoistDotOpEncV3(Operation* op) {
+  // Only consider custom conversions or arith ops.
+  if (!isa<FpToFpOp, BitcastOp>(op) && !isPureUnaryInlineAsm(op) &&
+      op->getDialect()->getTypeID() != TypeID::get<arith::ArithDialect>())
```
Here as well (`isa<>`).

chsigg commented on lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp:
```diff
@@ -15,6 +17,96 @@ namespace gpu {
+
+  // Don't hoist through u1 -> fp casts as they aren't supported in
+  // ElementwiseOpToLLVM::reorderValues().
+  if (isa<arith::UIToFPOp>(op)) {
```
Maybe combine these two conditions?

chsigg commented on lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp:
```diff
@@ -15,6 +17,96 @@ namespace gpu {
+  return true;
+}
+
+bool canHoistDotOpEncV3(Operation* op) {
```
Could you add a description (can refer to the function above)?

chsigg commented on lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp:
```diff
@@ -15,6 +17,96 @@ namespace gpu {
+  return true;
+}
+
+auto cloneSlice(PatternRewriter& rewriter, const SetVector<Operation *>& slice) {
```
I think this function could you use a description.

chsigg commented on lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp:
```diff
@@ -315,6 +380,150 @@ struct MMAV3UseRegOperand
+// MMAV3's analog of HoistLayoutConversion, for operand A only; will make WarpGroupDot
+// accept operand A in registers instead of shmem.
+//
+//  local_alloc(elementwise(x)) ->
```
I assume this describes the rewrite that is being performed? I think it could be a bit more accurate. I don't see the convert or the #dot_operand before the conversion, and it's not obvious what happened to the local_alloc.

chsigg commented on lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp:
```diff
@@ -315,6 +380,150 @@ struct MMAV3UseRegOperand
+    // Can only hoist operand 0
+    auto alloc = dotOp.getOperand(0).getDefiningOp<LocalAllocOp>();
+    if (!alloc || !alloc.getSrc())
+      return failure();
```
Prefer `rewriter.notifyMatchFailure()`.

vwbaker commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -435,15 +497,25 @@ assignMemoryLayouts(llvm::SmallVector<std::tuple<Operation *, int, Operation *>>
-      } else if (auto dot = dyn_cast<tt::DotOp>(use)) {
+      } else if (loadInfo.isMMAv3Registers || dot) {
+        // if warpGroupDot, we must now have operand A in registers since
+        // loadIsMMAv3Shared is false from above if-check
```
does this commend make sense anymore? it seems it's specifically in registers because we checked that it's specifically in registers, not just because it's not mmav3shared.

ggengnv replied:
I forgot to remove this, thanks for the reminder. Removed in https://github.com/openxla/triton/pull/19.

chsigg commented on lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp:
```diff
@@ -315,6 +380,150 @@ struct MMAV3UseRegOperand
+        return failure();
+
+      // We previously ensured that all ops in slice have at least one operand
+      for (auto operand : currOp->getOperands()) {
```
It's not easy for me to follow what this block of code checks. Please add a comment (and probably introduce the frontierOp term here instead of below).

vwbaker commented on third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/ConvertLayoutOpToLLVM.cpp:
```diff
@@ -88,11 +88,22 @@ struct LocalLoadOpConversion
+
+    if (mmaLayout.isHopper()) { // tensor core v3
+      assert(dotOperandLayout.getOpIdx() == 0);
+      res = SharedToDotOperandMMAv2::convertLayout(
```
rename to v2Or3 to avoid confusion please

vwbaker commented on third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMMAv2.cpp:
```diff
@@ -504,21 +528,50 @@ Type getSharedMemTy(Type argType) {
-      }
+      for (int k = 0; k < n1; ++k)
+        if (isHopper) {
+          // WGMMA.cpp expects different (m-major) ordering
```
Simply naming a file here is misleading as you could have just changed the expectation in that file. Why does it expect a different ordering? is this about the warp order being different on hopper than on ampere?

ggengnv replied:
Afaik m-major is the hardware-expected ordering for both Ampere and Hopper, and the reason for this difference is probably just code convention.

Ampere uses k-major ordering here, but in fact, there is [logic](https://github.com/triton-lang/triton/blob/main/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/DotOpToLLVM/MMAv2.cpp#L57) when lowering MMAv2 to LLVM to remap the ordering from k-major to m-major.

`WGMMA.cpp` has no analogous remapping logic. Before this PR, WGMMA can accept operand A as values in registers (i.e. LLVM struct) as produced [here](https://github.com/triton-lang/triton/blob/main/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/ConvertLayoutOpToLLVM.cpp#L617). This code lowers `convert_layout` from MMA (accumulator) encoding to DotOp encoding, which was for chained MMA's.

Therefore, if I was to change the expectation in `WGMMA.cpp`, I would also have to change the expectation for chained MMAs above. This modification, I expect, is probably a lot more effort compared to what I'm doing here, so I've opted to leave the current convention intact and just conditionally set the ordering here instead.

I can change the comment here to be clearer though.

vwbaker replied:
Thanks for the explanation! Just a comment is helpful and looks good now, thanks :)

vwbaker commented on third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMMAv2.cpp:
```diff
@@ -595,16 +651,18 @@ Value loadArg(ConversionPatternRewriter &rewriter, Location loc,
   auto shapePerCTA = getShapePerCTA(descTy);
   int bitwidth = descTy.getElementTypeBitWidth();
-  auto mmaLayout = mlir::cast<NvidiaMmaEncodingAttr>(encoding.getParent());
+  int mmaBitwidth = isHopper ? (32 / encoding.getKWidth()) : bitwidth;
```
can you add a comment why you are doing this?

vwbaker replied:
ah looks like it's mentioned below as "the sum of bitwidth of the elements in each quad should add up to 32.", but a short comment here would be nice too

vwbaker commented:
As a general comment, I think it would be nicer to review this & for openai to review it if this was separated into multiple PRs. You don't need to wait to have a performance improvement in every PR. It's enough to have a chain of easily-readable PRs that give a performance improvement in the end. This PR as a whole is doing a lot & could have at least separated  the enablement of registers in mmav3 & pipelining improvements for more performance gains

ggengnv commented:
> As a general comment, I think it would be nicer to review this & for openai to review it if this was separated into multiple PRs. You don't need to wait to have a performance improvement in every PR. It's enough to have a chain of easily-readable PRs that give a performance improvement in the end. This PR as a whole is doing a lot & could have at least separated the enablement of registers in mmav3 & pipelining improvements for more performance gains

I've split this into two PRs as you suggested :) - https://github.com/openxla/triton/pull/18 and https://github.com/openxla/triton/pull/19. I will be addressing the above comments in these new PRs from now on.

ggengnv commented:
Closing - split into https://github.com/openxla/triton/pull/18 and https://github.com/openxla/triton/pull/19

ggengnv commented:
Addressed all above comments in https://github.com/openxla/triton/pull/18 and https://github.com/openxla/triton/pull/19
